File: mcp/src/minecraft/femtocraft/power/tiles/TileEntityNanoCubePort.java
package femtocraft.power.tiles;

import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.world.World;
import net.minecraftforge.common.ForgeDirection;
import femtocraft.api.IInterfaceDevice;
import femtocraft.api.IPowerBlockContainer;
import femtocraft.core.multiblock.IMultiBlockComponent;
import femtocraft.core.multiblock.MultiBlockInfo;
import femtocraft.managers.research.EnumTechLevel;

public class TileEntityNanoCubePort extends TileEntityPowerBase implements
		IMultiBlockComponent {
	private static int storage = 500000;
	private MultiBlockInfo info;
	public boolean output;

	public TileEntityNanoCubePort() {
		super();
		info = new MultiBlockInfo();
		setMaxStorage(storage);
		setTechLevel(EnumTechLevel.NANO);
		output = false;
	}

	/*
	 * (non-Javadoc)
	 * 
	 * @see
	 * femtocraft.power.tiles.TileEntityPowerBase#canCharge(net.minecraftforge
	 * .common.ForgeDirection)
	 */
	@Override
	public boolean canCharge(ForgeDirection from) {
		if (!info.isValidMultiBlock() || output)
			return false;
		return super.canCharge(from);
	}

	/*
	 * (non-Javadoc)
	 * 
	 * @see
	 * femtocraft.power.tiles.TileEntityPowerBase#canAcceptPowerOfLevel(femtocraft
	 * .managers.research.EnumTechLevel,
	 * net.minecraftforge.common.ForgeDirection)
	 */
	@Override
	public boolean canAcceptPowerOfLevel(EnumTechLevel level,
			ForgeDirection from) {
		if (!info.isValidMultiBlock())
			return false;
		return super.canAcceptPowerOfLevel(level, from);
	}

	/*
	 * (non-Javadoc)
	 * 
	 * @see
	 * femtocraft.power.tiles.TileEntityPowerBase#canConnect(net.minecraftforge
	 * .common.ForgeDirection)
	 */
	@Override
	public boolean canConnect(ForgeDirection from) {
		if (!info.isValidMultiBlock())
			return false;
		return super.canConnect(from);
	}

	/*
	 * (non-Javadoc)
	 * 
	 * @see femtocraft.power.tiles.TileEntityPowerBase#getCurrentPower()
	 */
	@Override
	public int getCurrentPower() {
		if (info.isValidMultiBlock()) {
			if (isController())
				return super.getCurrentPower();

			IPowerBlockContainer fc = (IPowerBlockContainer) worldObj
					.getBlockTileEntity(info.x(), info.y(), info.z());
			if (fc != null) {
				return fc.getCurrentPower();
			}
		}
		return 0;
	}

	/*
	 * (non-Javadoc)
	 * 
	 * @see femtocraft.power.tiles.TileEntityPowerBase#getMaxPower()
	 */
	@Override
	public int getMaxPower() {
		if (info.isValidMultiBlock()) {
			if (isController())
				return super.getMaxPower();

			IPowerBlockContainer fc = (IPowerBlockContainer) worldObj
					.getBlockTileEntity(info.x(), info.y(), info.z());
			if (fc != null) {
				return fc.getMaxPower();
			}
		}
		return 0;
	}

	/*
	 * (non-Javadoc)
	 * 
	 * @see femtocraft.power.tiles.TileEntityPowerBase#getFillPercentage()
	 */
	@Override
	public float getFillPercentage() {
		if (info.isValidMultiBlock()) {
			if (isController())
				return super.getFillPercentage();

			IPowerBlockContainer fc = (IPowerBlockContainer) worldObj
					.getBlockTileEntity(info.x(), info.y(), info.z());
			if (fc != null) {
				return fc.getFillPercentage();
			}
		}
		return super.getFillPercentage();
	}

	/*
	 * (non-Javadoc)
	 * 
	 * @see
	 * femtocraft.power.tiles.TileEntityPowerBase#getFillPercentageForCharging
	 * (net.minecraftforge.common.ForgeDirection)
	 */
	@Override
	public float getFillPercentageForCharging(ForgeDirection from) {
		if (info.isValidMultiBlock()) {
			return output ? 1.f : 0.f;
		}
		return 1.f;
	}

	/*
	 * (non-Javadoc)
	 * 
	 * @see
	 * femtocraft.power.tiles.TileEntityPowerBase#getFillPercentageForOutput
	 * (net.minecraftforge.common.ForgeDirection)
	 */
	@Override
	public float getFillPercentageForOutput(ForgeDirection to) {
		if (info.isValidMultiBlock()) {
			return output ? 1.f : 0.f;
		}
		return 0.f;
	}

	/*
	 * (non-Javadoc)
	 * 
	 * @see
	 * femtocraft.power.tiles.TileEntityPowerBase#charge(net.minecraftforge.
	 * common.ForgeDirection, int)
	 */
	@Override
	public int charge(ForgeDirection from, int amount) {
		if (info.isValidMultiBlock() && !output) {
			if (isController())
				return super.charge(from, amount);

			IPowerBlockContainer fc = (IPowerBlockContainer) worldObj
					.getBlockTileEntity(info.x(), info.y(), info.z());
			if (fc != null) {
				return fc.charge(from, amount);
			}
		}
		return 0;
	}

	/*
	 * (non-Javadoc)
	 * 
	 * @see femtocraft.power.tiles.TileEntityPowerBase#consume(int)
	 */
	@Override
	public boolean consume(int amount) {
		if (info.isValidMultiBlock() && output) {
			if (isController())
				return super.consume(amount);

			IPowerBlockContainer fc = (IPowerBlockContainer) worldObj
					.getBlockTileEntity(info.x(), info.y(), info.z());
			if (fc != null) {
				return fc.consume(amount);
			}
		}
		return false;
	}

	private boolean isController() {
		return info.isValidMultiBlock()
				&& ((info.x() == xCoord) && (info.y() == yCoord) && (info.z() == zCoord));
	}

	/*
	 * (non-Javadoc)
	 * 
	 * @see
	 * femtocraft.core.tiles.TileEntityBase#onSideActivate(net.minecraft.entity
	 * .player.EntityPlayer, int)
	 */
	@Override
	public boolean onSideActivate(EntityPlayer par5EntityPlayer, int side) {
		if (isUseableByPlayer(par5EntityPlayer) && info.isValidMultiBlock()) {
			ItemStack item = par5EntityPlayer.getCurrentEquippedItem();
			if (item != null
					&& item.getItem() instanceof IInterfaceDevice
					&& ((IInterfaceDevice) item.getItem()).getInterfaceLevel().tier >= EnumTechLevel.NANO.tier) {
				output = !output;
				worldObj.markBlockForUpdate(xCoord, yCoord, zCoord);
				return true;
			}
		}
		return super.onSideActivate(par5EntityPlayer, side);
	}

	/*
	 * (non-Javadoc)
	 * 
	 * @see
	 * femtocraft.power.tiles.TileEntityPowerBase#readFromNBT(net.minecraft.
	 * nbt.NBTTagCompound)
	 */
	@Override
	public void readFromNBT(NBTTagCompound par1nbtTagCompound) {
		super.readFromNBT(par1nbtTagCompound);
		info.loadFromNBT(par1nbtTagCompound.getCompoundTag("info"));
		output = par1nbtTagCompound.getBoolean("output");
	}

	/*
	 * (non-Javadoc)
	 * 
	 * @see
	 * femtocraft.power.tiles.TileEntityPowerBase#writeToNBT(net.minecraft.nbt
	 * .NBTTagCompound)
	 */
	@Override
	public void writeToNBT(NBTTagCompound par1nbtTagCompound) {
		super.writeToNBT(par1nbtTagCompound);

		NBTTagCompound infoC = new NBTTagCompound();
		info.saveToNBT(infoC);
		par1nbtTagCompound.setTag("info", infoC);
		par1nbtTagCompound.setBoolean("output", output);
	}

	/*
	 * (non-Javadoc)
	 * 
	 * @see
	 * femtocraft.core.tiles.TileEntityBase#handleDescriptionNBT(net.minecraft
	 * .nbt.NBTTagCompound)
	 */
	@Override
	public void handleDescriptionNBT(NBTTagCompound compound) {
		super.handleDescriptionNBT(compound);
		info.loadFromNBT(compound.getCompoundTag("info"));
		output = compound.getBoolean("output");

		worldObj.markBlockForRenderUpdate(xCoord, yCoord, zCoord);
	}

	/*
	 * (non-Javadoc)
	 * 
	 * @see
	 * femtocraft.core.tiles.TileEntityBase#saveToDescriptionCompound(net.minecraft
	 * .nbt.NBTTagCompound)
	 */
	@Override
	public void saveToDescriptionCompound(NBTTagCompound compound) {
		super.saveToDescriptionCompound(compound);
		NBTTagCompound infoC = new NBTTagCompound();
		info.saveToNBT(infoC);
		compound.setTag("info", infoC);
		compound.setBoolean("output", output);
	}

	@Override
	public boolean hasGUI() {
		return info.isValidMultiBlock();
	}

	@Override
	public boolean isValidMultiBlock() {
		return info.isValidMultiBlock();
	}

	@Override
	public boolean formMultiBlock(World world, int x, int y, int z) {
		boolean result = info.formMultiBlock(world, x, y, z);
		worldObj.markBlockForUpdate(xCoord, yCoord, zCoord);
		worldObj.notifyBlocksOfNeighborChange(xCoord, yCoord, zCoord,
				worldObj.getBlockId(xCoord, yCoord, zCoord));
		return result;
	}

	@Override
	public boolean breakMultiBlock(World world, int x, int y, int z) {
		boolean result = info.breakMultiBlock(world, x, y, z);
		worldObj.markBlockForUpdate(xCoord, yCoord, zCoord);
		worldObj.notifyBlocksOfNeighborChange(xCoord, yCoord, zCoord,
				worldObj.getBlockId(xCoord, yCoord, zCoord));
		return result;
	}

	@Override
	public MultiBlockInfo getInfo() {
		return info;
	}
}

File: src/main/java/com/itszuvalex/femtocraft/power/tiles/TileEntityNanoCubePort.scala
/*
 * ******************************************************************************
 *  * Copyright (C) 2013  Christopher Harris (Itszuvalex)
 *  * Itszuvalex@gmail.com
 *  *
 *  * This program is free software; you can redistribute it and/or
 *  * modify it under the terms of the GNU General Public License
 *  * as published by the Free Software Foundation; either version 2
 *  * of the License, or (at your option) any later version.
 *  *
 *  * This program is distributed in the hope that it will be useful,
 *  * but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  * GNU General Public License for more details.
 *  *
 *  * You should have received a copy of the GNU General Public License
 *  * along with this program; if not, write to the Free Software
 *  * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 *  *****************************************************************************
 */
package com.itszuvalex.femtocraft.power.tiles

import com.itszuvalex.femtocraft.FemtocraftGuiConstants
import com.itszuvalex.femtocraft.api.core.{Configurable, Saveable}
import com.itszuvalex.femtocraft.api.power.{IPowerTileContainer, PowerContainer}
import com.itszuvalex.femtocraft.api.{EnumTechLevel, IInterfaceDevice}
import com.itszuvalex.femtocraft.core.traits.tile.MultiBlockComponent
import com.itszuvalex.femtocraft.power.tiles.TileEntityNanoCubePort._
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.nbt.NBTTagCompound
import net.minecraftforge.common.util.ForgeDirection

object TileEntityNanoCubePort {
  @Configurable(comment = "Maximum power storage.") private val storage   = 500000
  @Configurable(comment = "Tech level of power.") private   val techLevel = EnumTechLevel.NANO
}

@Configurable class TileEntityNanoCubePort extends TileEntityPowerBase with MultiBlockComponent {
  @Saveable(desc = true) var output = false

  override def defaultContainer: PowerContainer = new PowerContainer(techLevel, storage)

  override def consume(amount: Int): Boolean = {
    if (isValidMultiBlock && output) {
      if (isController) {
        return super.consume(amount)
      }
      val fc: TileEntityNanoCubePort = worldObj.getTileEntity(info.x, info.y, info.z).asInstanceOf[TileEntityNanoCubePort]
      if (fc != null) {
        return fc.controllerConsume(amount)
      }
    }
    false
  }

  private def controllerConsume(amount: Int) = isController && super.consume(amount)

  override def getFillPercentageForCharging(from: ForgeDirection): Float = {
    if (isValidMultiBlock) {
      return if (output) 1f else 0f
    }
    1f
  }

  override def charge(from: ForgeDirection, amount: Int): Int = {
    if (isValidMultiBlock && !output) {
      if (isController) {
        return super.charge(from, amount)
      }
      val fc: TileEntityNanoCubePort = worldObj.getTileEntity(info.x, info.y, info.z).asInstanceOf[TileEntityNanoCubePort]
      if (fc != null) {
        return fc.controllerCharge(from, amount)
      }
    }
    0
  }

  private def controllerCharge(from: ForgeDirection, amount: Int) = if (isController) super.charge(from, amount) else 0

  override def getMaxPower: Int = {
    if (isValidMultiBlock) {
      if (isController) {
        return storage
      }
      val fc = worldObj.getTileEntity(info.x, info.y, info.z).asInstanceOf[IPowerTileContainer]
      if (fc != null) {
        return fc.getMaxPower
      }
    }
    0
  }

  private def isController = isValidMultiBlock && ((info.x == xCoord) && (info.y == yCoord) && (info.z == zCoord))

  override def getFillPercentageForOutput(to: ForgeDirection): Float = {
    if (isValidMultiBlock) {
      return if (output) 1f else 0f
    }
    0f
  }

  override def canAcceptPowerOfLevel(level: EnumTechLevel, from: ForgeDirection) = isValidMultiBlock && super.canAcceptPowerOfLevel(level, from)

  override def getCurrentPower: Int = {
    if (isValidMultiBlock) {
      if (isController) {
        return super.getCurrentPower
      }
      val fc = worldObj.getTileEntity(info.x, info.y, info.z).asInstanceOf[IPowerTileContainer]
      if (fc != null) {
        return fc.getCurrentPower
      }
    }
    0
  }

  override def getFillPercentage: Float = {
    if (isValidMultiBlock) {
      if (isController) {
        return super.getFillPercentage
      }
      val fc: IPowerTileContainer = worldObj.getTileEntity(info.x, info.y, info.z).asInstanceOf[IPowerTileContainer]
      if (fc != null) {
        return fc.getFillPercentage
      }
    }
    super.getFillPercentage
  }

  override def canCharge(from: ForgeDirection) = !(!isValidMultiBlock || output) && super.canCharge(from)

  override def canConnect(from: ForgeDirection) = isValidMultiBlock && super.canConnect(from)

  override def handleDescriptionNBT(compound: NBTTagCompound) {
    super.handleDescriptionNBT(compound)
    setRenderUpdate()
  }

  override def onSideActivate(par5EntityPlayer: EntityPlayer, side: Int): Boolean = {
    if (canPlayerUse(par5EntityPlayer) && info.isValidMultiBlock) {
      val item = par5EntityPlayer.getCurrentEquippedItem
      if (item != null && item.getItem.isInstanceOf[IInterfaceDevice] && item.getItem.asInstanceOf[IInterfaceDevice].getInterfaceLevel.tier >= EnumTechLevel.NANO.tier) {
        output = !output
        worldObj.markBlockForUpdate(xCoord, yCoord, zCoord)
        return true
      }
      else {
        par5EntityPlayer.openGui(getMod, getGuiID, worldObj, info.x, info.y, info.z)
        return true
      }
    }
    false
  }

  override def getGuiID = FemtocraftGuiConstants.NanoCubeGuiID

  override def hasGUI = isValidMultiBlock
}
